applications/vanilla/controllers/api/DiscussionsApiController.php

```diff
@@ -645,4 +667,68 @@ public function put_bookmark($id, array $body) {
         $result = $this->discussionByID($id);
         return $out->validate($result);
     }
+
+    /**
+     * Set canonical url for a discussion.
+     *
+     * @param int $id The ID of the discussion.
+     * @param array $body The request body.
+     *       Ex: ["canonicalUrl" => "https://mydomain.com/some+path/"]
+     * @throws NotFoundException If unable to find the discussion.
+     * @return array
+     */
+    public function put_canonical($id, array $body) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $this->idParamSchema('in');
+        $in = $this->discussionPutCanonicalSchema('in')->setDescription('Set canonical url for a discussion.');
+        $out = $this->schema($this->discussionSchema(), 'out');
+
+        $body = $in->validate($body);
+
+        $row = $this->discussionByID($id);
+        if (!empty($row['Attributes']['CanonicalUrl'] ?? '')) {
+            throw new \Garden\Web\Exception\ClientException('Canonical url already set for this discussion.', 409);
+        };
+        $categoryID = $row['CategoryID'];
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->discussionModel->categoryPermission('Vanilla.Discussions.Edit', $categoryID);
+        }
+
+        $attributes = $row['Attributes'];
+        $attributes['CanonicalUrl'] = $body['canonicalUrl'];
```

Comment:
@initvector Do we have some in-built attributes handling in the models? I thought they were supposed to automatically serialize/unserialize these values.

Reply:
`DiscussionModel::getID` will handle unserializing the column on the way out. The column should be serialized with `DiscussionModel::save` on the way back in. However, `DiscussionModel::setProperty` doesn't perform any special field formatting and would require this manual re-serializing of the field contents. As an alternative, you might try something like:

```php
$this->discussionModel->save([
    "DiscussionID" => $id,
    "Attributes" => $attributes,
]);
```